cpp/src/sampling/negative_sampling_impl.cuh

```diff
@@ -412,16 +431,152 @@ std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> negativ
+
+      while (reduction > 0) {
+        size_t est_reduction_per_gpu = (reduction + comm_size - 1) / comm_size;
+        for (size_t i = 0; i < samples_per_gpu.size(); ++i) {
+          if (samples_per_gpu[i] > est_reduction_per_gpu) {
+            samples_per_gpu[i] -= est_reduction_per_gpu;
+            reduction -= est_reduction_per_gpu;
+          } else {
+            reduction -= samples_per_gpu[i];
+            samples_per_gpu[i] = 0;
+          }
+
+          if (reduction < est_reduction_per_gpu) est_reduction_per_gpu = reduction;
+        }
+      }
+      num_samples = samples_per_gpu[comm_rank];
+    }
+
+    // Mimic the logic of permute_range...
+    //
+    //  1) Randomly assign each entry to a GPU
+    //  2) Count how many are assigned to each GPU
+    //  3) Allgatherv (allgather?) to give each GPU a count for how many entries are destined for
+    //  that GPU 4) Identify extras/deficits for each GPU, arbitrarily adjust counts to make correct
+    //  5) Shuffle accordingly
+    //
+    rmm::device_uvector<int> gpu_assignment(srcs.size(), handle.get_stream());
+
+    cugraph::detail::uniform_random_fill(handle.get_stream(),
+                                         gpu_assignment.data(),
+                                         gpu_assignment.size(),
+                                         int{0},
+                                         int{comm_size},
+                                         rng_state);
+
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        gpu_assignment.begin(),
+                        gpu_assignment.end(),
+                        thrust::make_zip_iterator(srcs.begin(), dsts.begin()));
+
+    rmm::device_uvector<int> reduced_ranks(comm_size, handle.get_stream());
+    rmm::device_uvector<size_t> reduced_counts(comm_size, handle.get_stream());
+
+    reduced_ranks.resize(
+      thrust::distance(reduced_ranks.begin(),
+                       thrust::reduce_by_key(handle.get_thrust_policy(),
+                                             gpu_assignment.begin(),
+                                             gpu_assignment.end(),
+                                             thrust::make_constant_iterator(size_t{1}),
+                                             reduced_ranks.begin(),
+                                             reduced_counts.begin(),
+                                             thrust::equal_to<int>())
+                         .first),
+      handle.get_stream());
+    reduced_counts.resize(reduced_ranks.size(), handle.get_stream());
+
+    rmm::device_uvector<size_t> send_count(comm_size, handle.get_stream());
+    thrust::fill(handle.get_thrust_policy(), send_count.begin(), send_count.end(), 0);
+    thrust::scatter(handle.get_thrust_policy(),
+                    reduced_counts.begin(),
+                    reduced_counts.end(),
+                    reduced_ranks.begin(),
+                    send_count.begin());
```

Comment:
gpu_assignment is already sorted.

We can use thrust::upper_bound to find boundaries. This will require just comm_size binary searches and will be much faster.

Reply:
Fixed in the next push.